access/validator.go

```diff
@@ -70,28 +82,39 @@ type TransactionValidationOptions struct {
 	CheckScriptsParse            bool
 	MaxTransactionByteSize       uint64
 	MaxCollectionByteSize        uint64
+	CheckPayerBalance            bool
 }
 
 type TransactionValidator struct {
-	blocks                Blocks     // for looking up blocks to check transaction expiry
-	chain                 flow.Chain // for checking validity of addresses
-	options               TransactionValidationOptions
-	serviceAccountAddress flow.Address
-	limiter               RateLimiter
+	blocks                   Blocks     // for looking up blocks to check transaction expiry
+	chain                    flow.Chain // for checking validity of addresses
+	options                  TransactionValidationOptions
+	serviceAccountAddress    flow.Address
+	limiter                  RateLimiter
+	scriptExecutor           execution.ScriptExecutor
+	verifyPayerBalanceScript []byte
 }
 
 func NewTransactionValidator(
 	blocks Blocks,
 	chain flow.Chain,
 	options TransactionValidationOptions,
-) *TransactionValidator {
-	return &TransactionValidator{
-		blocks:                blocks,
-		chain:                 chain,
-		options:               options,
-		serviceAccountAddress: chain.ServiceAddress(),
-		limiter:               NewNoopLimiter(),
+	executor execution.ScriptExecutor,
+) (*TransactionValidator, error) {
+	env := systemcontracts.SystemContractsForChain(chain.ChainID()).AsTemplateEnv()
+	if options.CheckPayerBalance && executor == nil {
+		return nil, errors.New("transaction validator cannot use checkPayerBalance with nil executor")
 	}
```

Comment:
nit: instantiate after the check
```suggestion
	if options.CheckPayerBalance && executor == nil {
		return nil, errors.New("transaction validator cannot use checkPayerBalance with nil executor")
	}
	env := systemcontracts.SystemContractsForChain(chain.ChainID()).AsTemplateEnv()
```